File: solutions/fortran/sieve/1/sieve.f90
module sieve
  implicit none

contains

  function primes(limit) result(array)
    integer, intent(in) :: limit
    integer, allocatable :: array(:)
    logical :: is_prime(limit)
    integer :: i, j
    is_prime = .true.

    do i = 2, int(sqrt(real(limit)))
      if (is_prime(i)) then
        is_prime([(j, j = i**2, limit, i)]) = .false.
      end if
    end do

    array = pack([(i,i=2, limit)], is_prime(2:))
  end function primes

end module sieve

File: solutions/csharp/sieve/1/Sieve.cs
using System;
using System.Collections.Generic;

public static class Sieve
{
    public static int[] Primes(int limit)
    {
        if (limit < 2) throw new ArgumentOutOfRangeException();
        var isPrime = new bool[limit + 1];
        for (int i = 2; i <= limit; i++) isPrime[i] = true;

        // cross out multiples
        for (var i = 2; i <= limit; i++)
        {
            if (!isPrime[i]) continue;

            // knock out multiples of i.
            for(var j = i * 2; j <= limit; j+= i)
                isPrime[j] = false;
        }

        var ret = new List<int>();
        for(var i = 2; i <= limit; i++)
          if(isPrime[i]) ret.Add(i);
        return ret.ToArray();
    }
}